### PR TITLE
feat(wallet): Portfolio Filters and Display Settings

### PR DESCRIPTION
**Pull request by Douglashdaniel**
## Description 
Introduces a `Portfolio Filters and Display Settings` modal.

Settings include:
- `Sort assets` dropdown (High to Low, Low to High, A to Z, Z to A)
- `Hide small balances` toggle (will hide balance with a value less than $1)
- `Select networks` filter (Assets of checked networks will be displayed in portfolio, note: by default all `test networks` will be unchecked, if a user checks a `test network` the value will be included in the `portfolio`.)
- `Select accounts` filter (Balances will be reflected by what `accounts` are checked.)

The `Edit Visible Assets` button is no longer at the bottom of the page but is now a new button to the left of the `Portfolio Filters and Display Settings` button at the top right of the portfolio assets list.

Resolves <https://github.com/brave/brave-browser/issues/30496>

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### Testing `Sort assets`:
1. Go to the `Portfolio` page and open the `Filters and display settings`
2. Change the `Sort assets` option and click `Save changes`
3. The portfolio should reflect your selected option
4. Test this for all options.

https://github.com/brave/brave-core/assets/40611140/b470e8a3-ee75-44ec-9dc2-562dd368956e

### Testing `Hide small balances`:
1. Go to the `Portfolio` page and open the `Filters and display settings`
2. Toggle on the `Hide small balances` toggle and click `Save changes`
3. The portfolio should hide assets with a balance less than $1

https://github.com/brave/brave-core/assets/40611140/f2ac399d-f3e5-4848-8279-36e4b9bbf2e5

### Testing `Select networks`:
1. Go to the `Portfolio` page and open the `Filters and display settings`
2. Check or Uncheck `networks` and click `Save changes`
3. The portfolio should reflect your `Selected networks` (note: by default all `test networks` will be unchecked, if a user checks a `test network` the value will be included in the `portfolio`).
4. Test the `Select all` and `Deselect all` networks button.

https://github.com/brave/brave-core/assets/40611140/dc9947f3-0269-4dd6-89af-02941a498041

### Testing `Select accounts`:
1. Go to the `Portfolio` page and open the `Filters and display settings`
2. Check or Uncheck `accounts` and click `Save changes`
3. The portfolio should reflect your `Selected accounts` (note: `Selected Networks` takes a higher priority and assets will remain visible but will show a `0` balance if associated accounts are hidden.)
4. Test the `Select all` and `Deselect all` accounts button.

https://github.com/brave/brave-core/assets/40611140/f4a05c2e-c201-45c1-b5ed-340b5cea2055

### CODE REVIEW

File: components/resources/wallet_strings.grdp [hunk 1]
```diff
@@ -563,10 +564,24 @@
...
   <message name="IDS_BRAVE_WALLET_NETWORK_FILTER_ALL" desc="Network Filter Selector All Networks">All Networks</message>
-  <message name="IDS_BRAVE_WALLET_ASSET_FILTER_LOW_TO_HIGH" desc="Aseet Filter Selector Low to High">Balance ↑</message>
-  <message name="IDS_BRAVE_WALLET_ASSET_FILTER_HIGH_TO_LOW" desc="Aseet Filter Selector High to Low">Balance ↓</message>
-  <message name="IDS_BRAVE_WALLET_ASSET_FILTER_A_TO_Z" desc="Aseet Filter Selector A to Z">Name ↓</message>
-  <message name="IDS_BRAVE_WALLET_ASSET_FILTER_Z_TO_A" desc="Aseet Filter Selector Z to A">Name ↑</message>
+  <message name="IDS_BRAVE_WALLET_ASSET_FILTER_LOW_TO_HIGH" desc="Aseet Filter Selector Low to High">Low to High</message>
+  <message name="IDS_BRAVE_WALLET_ASSET_FILTER_HIGH_TO_LOW" desc="Aseet Filter Selector High to Low">High to Low</message>
+  <message name="IDS_BRAVE_WALLET_ASSET_FILTER_A_TO_Z" desc="Aseet Filter Selector A to Z">A to Z</message>
+  <message name="IDS_BRAVE_WALLET_ASSET_FILTER_Z_TO_A" desc="Aseet Filter Selector Z to A">Z to A</message>
```
Comment (mkarolin) [inline components/resources/wallet_strings.grdp:570]: Typo in the `desc` attribute: `Aseet` -> `Asset`

File: components/resources/wallet_strings.grdp [hunk 2]
```diff
@@ -563,10 +564,24 @@
...
+  <message name="IDS_BRAVE_WALLET_PORTFOLIO_FILTERS_TITLE" desc="Portfolio Filters and Display settings modal">Filters and display settings</message>
+  <message name="IDS_BRAVE_WALLET_SORT_ASSETS" desc="Portfolio settings sort assets dropdown title">Sort assets</message>
+  <message name="IDS_BRAVE_WALLET_SORT_ASSETS_DESCRIPTION" desc="Portfolio settings sort assets dropdown description">Sort by fiat value or name</message>
+  <message name="IDS_BRAVE_WALLET_SMALL_BALANCES" desc="Portfolio settings hide small balances toggle title">Hide small balances</message>
+  <message name="IDS_BRAVE_WALLET_SMALL_BALANCES_DESCRIPTION" desc="Portfolio settings hide small balances toggle description">Assets with value less than <ph name="FIAT_VALUE"><ex>Ethereum</ex>$1</ph></message>
```
Comment (mkarolin) [inline components/resources/wallet_strings.grdp:575]: The `<ex>` value doesn't seem to match the implied numeric type.

File: components/resources/wallet_strings.grdp (hunk 1 above)
Comment (Douglashdaniel) [reply components/resources/wallet_strings.grdp]: Fixed!

**Review by onyb — APPROVED** [review]
LGTM 🚀

File: components/brave_wallet_ui/utils/network-utils.ts [hunk 3]
```diff
@@ -117,3 +117,11 @@ export const reduceNetworkDisplayName = (name?: string) => {
...
   return name.split(' ')[0]
 }
+
+export const getNetworkIdKey =
+  (
+    chainId: string,
+    coin: BraveWallet.CoinType
+  ) => {
+    return `${chainId}-${coin}`
+  }
```
[the remark this answers is not in the document]
Comment (Douglashdaniel) [reply components/brave_wallet_ui/utils/network-utils.ts]: Good idea! Updated!

File: components/resources/wallet_strings.grdp [hunk 4]
```diff
@@ -563,10 +564,24 @@
   <message name="IDS_BRAVE_WALLET_BUY_SARDINE_DESCRIPTION" desc="Sardine description in buy options list">Easiest, fastest and cheapest way to buy crypto with card and bank transfers.</message>
   <message name="IDS_BRAVE_WALLET_BUY_TRANSAK_DESCRIPTION" desc="Transak description in buy options list">Instant buy with your bank account. Lower fees.</message>
   <message name="IDS_BRAVE_WALLET_NETWORK_FILTER_ALL" desc="Network Filter Selector All Networks">All Networks</message>
-  <message name="IDS_BRAVE_WALLET_ASSET_FILTER_LOW_TO_HIGH" desc="Aseet Filter Selector Low to High">Balance ↑</message>
-  <message name="IDS_BRAVE_WALLET_ASSET_FILTER_HIGH_TO_LOW" desc="Aseet Filter Selector High to Low">Balance ↓</message>
-  <message name="IDS_BRAVE_WALLET_ASSET_FILTER_A_TO_Z" desc="Aseet Filter Selector A to Z">Name ↓</message>
-  <message name="IDS_BRAVE_WALLET_ASSET_FILTER_Z_TO_A" desc="Aseet Filter Selector Z to A">Name ↑</message>
+  <message name="IDS_BRAVE_WALLET_ASSET_FILTER_LOW_TO_HIGH" desc="Aseet Filter Selector Low to High">Low to High</message>
+  <message name="IDS_BRAVE_WALLET_ASSET_FILTER_HIGH_TO_LOW" desc="Aseet Filter Selector High to Low">High to Low</message>
```
[the remark this answers is not in the document]
Comment (Douglashdaniel) [reply components/resources/wallet_strings.grdp]: 🤦 totally overlooked those 2.. they are fixed now! :)

**Review by mkarolin — APPROVED** [review]
`strings`++

File: components/brave_wallet_ui/components/desktop/popup-modals/filter-modals/filter-components/filter-dropdown-section.tsx [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+  namespace JSX {
+    interface IntrinsicElements {
+      'dropdown-option': React.HTMLAttributes<HTMLElement> & {
+        value?: string
+        children?: any
+        key?: React.Key
+      }
+    }
+  }
+}
```
Comment (josheleonard) [inline components/brave_wallet_ui/components/desktop/popup-modals/filter-modals/filter-components/filter-dropdown-section.tsx:51]: please move this to a declarations file.

maybe `declarations/elements.d.ts`?

**Review by josheleonard — CHANGES_REQUESTED** [review]

File: components/brave_wallet_ui/utils/network-utils.ts [hunk 6]
```diff
@@ -117,3 +117,10 @@ export const reduceNetworkDisplayName = (name?: string) => {
   }
   return name.split(' ')[0]
 }
+
+export const getNetworkIdKey =
```
[the remark this answers is not in the document]
Comment (Douglashdaniel) [reply components/brave_wallet_ui/utils/network-utils.ts]: Switched to use `networkEntityAdapter.selectId` :)

File: components/brave_wallet_ui/components/desktop/popup-modals/filter-modals/filter-components/filter-dropdown-section.tsx (hunk 5 above)
Comment (Douglashdaniel) [reply components/brave_wallet_ui/components/desktop/popup-modals/filter-modals/filter-components/filter-dropdown-section.tsx:54]: Added an issue to remove this and use the provided `leo-option` once a `key` prop is added.
https://github.com/brave/brave-browser/issues/30826

**Review by SergeyZhukovsky — APPROVED** [review]
++ core part

**Review by josheleonard — APPROVED** [review]
++

**Pull request merged**